src/core/annotation.js

```diff
@@ -920,37 +933,141 @@ class TextWidgetAnnotation extends WidgetAnnotation {
+        operatorList.addOp(OPS.beginAnnotation, [
+          this.data.rect,
+          transform,
+          matrix,
+        ]);
+
+        const stream = new StringStream(content);
+        return evaluator
+          .getOperatorList({
+            stream,
+            task,
+            resources: this.fieldResources,
+            operatorList,
+          })
+          .then(function () {
+            operatorList.addOp(OPS.endAnnotation, []);
+            return operatorList;
+          });
+      }
+    );
+  }
+
+  async getAppearance(evaluator, task, annotationStorage) {
+    if (!annotationStorage) {
+      return null;
+    }
+    const value = annotationStorage[this.data.id] || "";
+    if (value === "") {
+      return null;
+    }
+
+    const borderWidth = this.data.borderStyle.width;
+
+    // Default horizontal padding: can we have an heuristic to guess it?
+    const hPadding = borderWidth + 2;
+
+    // TODO: Handle the case where the font is not defined:
+    // fallback on helv maybe?
+    // TODO: Handle the case where fontSize is null
+    const [font, fontSize] = await this.getFontData(evaluator, task);
+
+    const vPadding = borderWidth + Math.ceil(Math.abs(font.descent) * fontSize);
+    const defaultAppearance = this.data.defaultAppearance;
+    const totalWidth = this.data.rect[2] - this.data.rect[0];
+
+    if (this.data.comb) {
+      const combWidth = (totalWidth / this.data.maxLen).toFixed(2);
+      let buf = `/Tx BMC q BT ${defaultAppearance} 1 0 0 1 ${hPadding} ${vPadding} Tm`;
+      let first = true;
+      for (const character of value) {
+        if (first) {
+          buf += ` (${character}) Tj`;
+          first = false;
+        } else {
+          buf += ` ${combWidth} 0 Td (${character}) Tj`;
+        }
+      }
+      buf += " ET Q EMC";
+      return buf;
+    }
+
+    const alignment = this.data.textAlignment;
+    if (alignment === 0 || alignment > 2) {
+      // Left alignment: nothing to do
+      return `/Tx BMC q BT ${defaultAppearance} 1 0 0 1 ${hPadding} ${vPadding} Tm (${value}) Tj ET Q EMC`;
     }
 
+    // We need to get the width of the text in order to align it correctly
+    const glyphs = font.charsToGlyphs(value);
+
+    const scale = fontSize / 1000;
+    let width = 0;
+    for (const glyph of glyphs) {
+      width += glyph.width * scale;
+    }
+
+    let shift;
+    if (alignment === 1) {
+      // Center
+      shift = (totalWidth - width) / 2;
+    } else {
+      // Right
+      shift = totalWidth - width - hPadding;
+    }
+
+    return `/Tx BMC q BT ${defaultAppearance} 1 0 0 1 ${shift} ${vPadding} Tm (${value}) Tj ET Q EMC`;
+  }
+
+  async getFontData(evaluator, task) {
     const operatorList = new OperatorList();
+    const initialState = new EvalState();
+    await evaluator.getOperatorList({
+      stream: new StringStream(this.data.defaultAppearance),
+      task,
+      resources: this.fieldResources,
+      operatorList,
+      initialState,
```

Comment:
Please note the default value in https://github.com/mozilla/pdf.js/blob/72d71ba6a556093a1c88525e827e59590c9151f1/src/core/evaluator.js#L1224-L1235 hence you shouldn't need to pass the `initialState` here (nor export `EvalState` in the other file).

Reply:
I removed EvalState export but I want to pass an initialState to be able to get the font and the font size from it in order to compute text width (and split it for multiline textfields).

Reply:
If you want to chat about that or anything else, then I'm on riot in the pdf.js channel, my nick is "calixte" (in CEST timezone).

Reply:
Sorry, I completely overlooked that you actually accessed the state afterwards; in that case exporting/using it like you previously did is probably OK.

(Although it does feel slightly "hacky" to do things in that way, given how the `getOperatorList` method in currently implemented/used, however at the top of my head I don't really have a much better suggestion.)